src/components/DataTable.tsx

```diff
@@ -0,0 +1,225 @@
+import React, { useState, useRef } from 'react';
+import {
+  Button,
+  Paper,
+  Card,
+  Typography,
+  Table,
+  TableBody,
+  TableCell,
+  TableContainer,
+  TableHead,
+  TableRow,
+  TablePagination,
+  Collapse,
+  useTheme,
+} from '@mui/material';
+import { CloudUpload, ExpandMore, ExpandLess } from '@mui/icons-material';
+import { v4 as uuidv4 } from 'uuid';
+import useStore from '../stores/store';
+
+function DataTable() {
+  // component state
+  const [TSVFile, setTSVFile] = useState<File | null>(null);
+  const [data, setData] = useState<string[][]>([]);
+  const [headers, setHeaders] = useState<string[]>([]);
+  const [page, setPage] = useState(0);
+  const [rowsPerPage, setRowsPerPage] = useState(5);
+  const [isPreviewOpen, setIsPreviewOpen] = useState(false);
+
+  const fileInputRef = useRef<HTMLInputElement>(null);
+  const theme = useTheme();
+
+  // store state
+  const setDataTable = useStore((state) => state.setDataTable);
+  const initializeColumns = useStore((state) => state.initializeColumns);
+
+  const transformDataToColumns = (rows: string[][]) => {
+    const columnData: { [key: number]: string[] } = {};
+    if (rows.length > 0) {
+      const numColumns = rows[0].length;
+      for (let col = 0; col < numColumns; col += 1) {
+        columnData[col + 1] = rows.map((row) => row[col]);
+      }
+    }
+    return columnData;
+  };
+
+  const transformHeadersToColumns = (tsvHeaders: string[]) => {
+    const columns: { [key: number]: { header: string } } = {};
+    tsvHeaders.forEach((header, index) => {
+      columns[index + 1] = { header };
+    });
+    return columns;
+  };
+
+  const processFile = (file: File) => {
+    const reader = new FileReader();
+    reader.onload = (e) => {
+      const content = e.target?.result as string;
+      const rows = content.split('\n').map((row) => row.split('\t'));
+      setHeaders(rows[0]);
+      setData(rows.slice(1));
+
+      const columnData = transformDataToColumns(rows.slice(1));
+      setDataTable(columnData);
+
+      const columnHeaders = transformHeadersToColumns(rows[0]);
+      initializeColumns(columnHeaders);
+    };
+    reader.readAsText(file);
+  };
+
```

Comment:
All of these methods I suggest we put in the store or into custom hooks, even if we don't reuse them yet. one, it'll make them much easier to test and that in itself should help mock them in integration tests, and two, it'll make the component so much easier to read